tests/cpu/native_backtrace/Makefile

```diff
@@ -5,3 +5,13 @@ USEMODULE += backtrace
 BOARD_WHITELIST := native
 
 include $(RIOTBASE)/Makefile.include
+
+ifeq ($(BOARD),native)
+  ifneq (,$(filter arch_64bit,$(FEATURES_USED)))
+    # 64 bit
+    CFLAGS += -DBACKTRACE_SIZE=3
+  else
+    # 32 bit
+    CFLAGS += -DBACKTRACE_SIZE=4
+  endif
```

Comment:
This I don't understand - why does 64 bit only need a backtrace size of 3?

Reply:
To be honest, I am not entirely sure.
It is related in some way to the combination of `makecontext()` and `backtrace()`.

`native` `backtrace()` result:
```
0x08049641: main+0x3B
0x08049cef: main_trampoline+0x36
0xf7cc57d9: makecontext+0x109
0x08049bf8: thread_create+0x185
```

`native64` `backtrace()` result:
```
0x0000004015e7: main+0x31
0x000000401ccf: main_trampoline+0x31
0x7f15d56cd130: libc.so.6(+0x5a130)
```

The gdb backtrace command, on the other hand, gives the same result
```
native gdb backtrace result:
#0  0x080496d9 in backtrace_print () at <RIOT>/cpu/native/backtrace/backtrace.c:23
#1  0x08049641 in main () at <RIOT>/tests/cpu/native_backtrace/main.c:28

native64 gdb backtrace result:
#0  0x000000000040167e in backtrace_print () at <RIOT>/cpu/native/backtrace/backtrace.c:23
#1  0x00000000004015e7 in main () at <RIOT>/tests/cpu/native_backtrace/main.c:28
```

Reply:
I did some more debugging and tried to figure out why exactly the backtraces are different.
The `native_backtrace` test only checks if the returned backtrace is at least the given length.
But at least on my machine, the `x86` backtrace actually returns a slightly longer backtrace (size 5 instead of 4), and I tried to reconstruct how the `backtrace()` function unwinds the stack.

Frame | Memory location | Backtrace 
--|-------------------------------|----------------------------
1 | 0x08055e7c <main_stack+11740> | 0x08049631 0x8049631 <main+59>
2 | 0x08055ebc <main_stack+11804> | 0x080497ec <main_trampoline+54>
3 | 0x08055eec <main_stack+11852> | 0xf7dc37d9 <makecontext+265>
4 | 0xffffcddc                    | 0x08049e12 <thread_create+389>
5 | 0xffffce6c                    | 0x080498c0 <cpu_switch_context_exit+47>
6 | 0x00000000                    |

After frame 3, `backtrace()` leaves the main stack and walks through dynamically allocated memory, which is basically the internals of the makecontext function and results in a wrong backtrace.
`backtrace()` is not really designed to handle this, because the user context functions use a lot of assembler and kernel calls to implement their logic instead of simple function calls.
The result is that `backtrace()` more or less just randomly walks through memory.
In addition, the user context implementations in glibc are quite different for i386 and x86_64, which is the reason for the different backtraces.

Since the backtrace is wrong after frame 3, my suggestion would be to set `CFLAGS += -DBACKTRACE_SIZE=3` for all architectures.